tests/test_x86.py

```diff
@@ -428,3 +428,161 @@ def sgemm_full(
+        for i in range(N):
+            if x[i] < v[i]:
+                expected[i] = y[i]
+            else:
+                expected[i] = z[i]
+
+        out = np.array(np.random.rand(N), dtype=np.float32)
+        x_copy = x.copy()
+        v_copy = v.copy()
+        y_copy = y.copy()
+        z_copy = z.copy()
+
+        fn(None, N, out, x, v, y, z)
+
+        np.testing.assert_almost_equal(out, expected)
+        np.testing.assert_almost_equal(x, x_copy)
+        np.testing.assert_almost_equal(v, v_copy)
+        np.testing.assert_almost_equal(y, y_copy)
+        np.testing.assert_almost_equal(z, z_copy)
+
+    N = 50
+    x = np.array([float(i) for i in range(N)], dtype=np.float32)
+    v = np.array(
+        [1, 0, 3, 4, 1, 1, 9, 0]
+        + [100] * 8
+        + [0] * 8
+        + [float(i) for i in range(24, N)],
+        dtype=np.float32,
+    )
+    y = np.array([float(i) for i in range(200, 200 + N)], dtype=np.float32)
+    z = np.array([float(i) for i in range(300, 300 + N)], dtype=np.float32)
+
+    # Correctness testing
+    run_and_check(N, x, v, y, z)
+
+    # Precision testing
+    N = 111
+    run_and_check(
+        N,
+        np.array(np.random.rand(N), dtype=np.float32),
+        np.array(np.random.rand(N), dtype=np.float32),
+        np.array(np.random.rand(N), dtype=np.float32),
+        np.array(np.random.rand(N), dtype=np.float32),
+    )
+
+
+@pytest.mark.isa("AVX2")
+def test_avx2_accumulate_assume_associative(compiler):
+    @proc
+    def accumulate_buffer(x: f32[8], result: [f32][1]):
+        tmp_result: f32
+        tmp_result = result[0]
+        for i in seq(0, 8):
+            tmp_result += x[i]
+        result[0] = tmp_result
+
+    accumulate_buffer = stage_mem(accumulate_buffer, "for i in _:_", "x[0:8]", "xReg")
+    accumulate_buffer = set_memory(accumulate_buffer, "xReg", AVX2)
+    accumulate_buffer = replace(accumulate_buffer, "for i0 in _:_", mm256_loadu_ps)
+    accumulate_buffer = replace(
+        accumulate_buffer, "for i in _:_", avx2_accumulate_assume_associative
+    )
+    accumulate_buffer = simplify(accumulate_buffer)
+
+    fn = compiler.compile(
+        accumulate_buffer, skip_on_fail=True, CMAKE_C_FLAGS="-march=skylake"
+    )
+
+    def run_and_check(x, result):
+        expected = result.copy()
+        for i in range(8):
+            expected[0] += x[i]
+
+        x_copy = x.copy()
+
+        fn(None, x, result)
+
+        # lower precision checking because we are assuming float addition is associative in the instruction
+        np.testing.assert_almost_equal(result, expected, decimal=4)
+        np.testing.assert_almost_equal(x, x_copy)
+
+    result = np.array([0.0], dtype=np.float32)
+    x = np.array([1, 2, 3, 1, 2, 3, 7, 2], dtype=np.float32)
+
+    fn(None, x, result)
+
+    run_and_check(x, result)
+    run_and_check(
+        np.array(np.random.rand(8), dtype=np.float32),
+        np.array([np.random.rand()], dtype=np.float32),
+    )
```

Comment:
This test may not be stable since floating point addition is calculated with different associativity. Are 4 decimals of accuracy stable enough?

Reply:
Just don't make it random at runtime. Pick some specific values, even if at random once.

Reply:
Done